src/Core.hs

```diff
@@ -108,6 +111,9 @@ processOps core = go Null
           atomically $ do
             writeTVar (coreCurrentValue core) newValue
             writeTBQueue (coreUpdates core) (Just $ Updated (opPath op) newValue)
+          -- persist the updated Json object to disk
+          -- TODO: make it configurable how often we do this (like in Redis)
+          writeFile "icepeak.json" (encodeToLazyText newValue)
```

Comment:
One small thing we can do here that would greatly increase reliability, is to write it to `icepeak.json.new`, and afterwards rename `icepeak.json.new` to `icepeak.json`. Rename is atomic on Posix (see `man 3 rename`), which guarantees that `icepeak.json` is a complete json file.\* Currently, if the application is killed or crashes while writing the file, the file on disk will be incomplete, and invalid json, which will fail to recover at startup. By doing the rename, it will load a slightly older version, but at least it will be able to load *something*.

<sub>\* There is still the complication of OS page caches, file system drivers, disk drivers, virtual machine hosts, raid controllers, and disk firmware buffering writes. Actually making sure that data is on disk is a Hard Problem that we probably don’t want to worry about at this point. But killing the app while it is in the middle of a write is somewhat common.</sub>

Also, this should take the filename from the config.

We might also do the writing in a separate thread with a queue between them in the future. (Trading latency for the possibility of losing the last few writes.) From what I’ve seen in other applications, when the json gets big enough (which could be the case here, even if every individual write is small), serializing json dominates execution time. With a bounded queue, handling a sustained load of puts would still block when the queue is full, but when there is a temporary spike of writes, processing could continue with little latency impact. The size of the queue is quite literally the knob to tune, as it is the number of writes that you are willing to lose. The writer thread could also take shortcuts, dropping values if there are more waiting behind them in the queue anyway. That would increase write throughput, while still putting an upper bound on the amount of writes that you are willing to lose (twice the length of the queue).

Reply:
> Also, this should take the filename from the config.

I added the config to the core, since we will probably need more relevant settings from it in the future anyway.

> One small thing we can do here that would greatly increase reliability, is to write it to icepeak.json.new, and afterwards rename icepeak.json.new to icepeak.json.

Good idea, I added that.

> We might also do the writing in a separate thread with a queue between them in the future.

Yes, I was thinking about that as well - it would be nice to make this configurable by they user (similar to how Redis does it). Then users can decide themselves whether they value robustness or performance more.